### ISSUE TITLE
Commit 0f9b3cb broke Quick Explorer keyboard nav w/preview

### ISSUE THREAD

**Issue by pjeby** [nothingislost/obsidian-hover-editor#125]
The keyup handler detects the keyup of the arrow key and cancels the popover.

I can work around this on the QE side by forcing the abort controller to unload in a subsequent microtick, but the keyup handler assumes way too much here.  The truth is that even the default page preview plugin can "inadvertently" pop things up if you hold ctrl to do another hotkey; the only difference with HE is that autofocus means your hotkey could target the wrong pane. The keyup handler doesn't necessarily address that problem though, since the command would trigger on keydown, not keyup.

In short, I think the commit should be reverted: it's not really HE's job to watch out for people placing their cursor on links; at most it could be changed to disable autofocus for the popup in question, which would prevent the popup from stealing focus.  In QE's case this wouldn't cause any problems since it keeps focus and key scope anyway.  It should possibly also be watching keydown instead of keyup.

**Comment by nothingislost** [comment]
here's my proposed fix.  the main addition is the `isModShortcut` check

```js
    // to prevent mod based keyboard shortcuts from accidentally triggering popovers
    const onKeyDown = function (event: KeyboardEvent) {
      if (!editor) return;
      const modKey = Platform.isMacOS ? "Meta" : "Control";
      const isModShortcut = Platform.isMacOS ? event.metaKey : event.ctrlKey;
      if (!editor.onHover && editor.state !== PopoverState.Shown && isModShortcut && event.key !== modKey) {
        editor.state = PopoverState.Hidden;
        editor.hide();
        editor.lockedOut = true;
        setTimeout(unlock, 1000);
      } else {
        document.body.removeEventListener("keydown", onKeyDown, true);
      }
    };

    document.body.addEventListener("mousedown", onMouseDown, true);
    document.body.addEventListener("keydown", onKeyDown, true);
    controller.register(() => {
      document.body.removeEventListener("mousedown", onMouseDown, true);
      document.body.removeEventListener("keydown", onKeyDown, true);
    });
```

**Comment by pjeby** [comment]
I'm not sure the isModShortcut is even needed, since if the goal here is to avoid popups stealing focus, that can happen with regular keys, alt keys, etc.

Honestly, I'm still confused about the intended use case here, since the core page preview plugin doesn't do anything like this, and the only difference with HE is the possibility of stealing focus, which could be handled by suppressing autofocus for the popup rather than cancelling its loading process.

I'm also worried about the effective API, in the sense that this entire thing makes programmatic interaction with popups harder -- QE is monkeying the crap out of this to work around it.  It would be less of an issue if all it did was turn off autofocus instead of canceling the popup.

Basically, figuring out popover state is a royal mess to begin with, without adding this to the state machine.  If I ever add any mod-something shortcuts to QE, they're going to possibly interact with this in a negative way when used between selecting the item and the popup's appearance.

Cancelling autofocus is less problematic, OTOH, because there aren't many use cases for trying to send commands to a popover before it exists.  And without autofocus, the negative impact of an excess popover seems no more than it was with the core page preview plugin.

Unless I'm missing something about the use case that prompted this change?

**Comment by nothingislost** [comment]
See https://github.com/nothingislost/obsidian-hover-editor/issues/106

The idea is less about focus and more about not showing the popover at all if a mod based keyboard shortcut was pressed. A single mod key press should be the only thing that can invoke a popover, not a compound mod key shortcut.

The problem is, you don't know if a mod key based shortcut is about to be invoked until a few milliseconds after the initial mod key press (since people press mod first, then the second key). This is why the handler currently only lives during the "showing" state of the popover as it uses that brief period of time to look for a compound shortcut.

Just to be clear, I'm not protective of this code and agree that it's not ideal in its current state. I probably rushed the implementation of this a bit but I do want to try and address the issue somehow instead of removing it completely.

I haven't been able to brainstorm much on better ways to accomplish this but one option might be to just provide a customizable trigger key so that people can use something other than mod.

**Comment by pjeby** [comment]
It's kind of funny how different workflows can be; for me, most of my page preview toggles are set to *not* require Ctrl, so the moment my mouse is over something I don't want it over, *I move it* to get rid of the popup. :grin:

Anyway, the customizable trigger won't work for plugins like Calendar, nor any other plugin that does its own logic for that.

I guess this version is at least *better*, but I'm still honestly confused as to how these folks didn't have problems with the original Page Preview.  (Though perhaps it's a matter of size, since the PP default is rather small.)

Maybe if there's an explicit API to bypass this (vs. abortController.unload()) that QE could use, that would at least allow QE to ignore any issues with this, and not have to worry about unloading other stuff that might actually be needed.  For example, if the editor had a `.shouldCancelOnKey` property defaulting to true, QE could set it false when the parent is assigned and the handler could ignore keystrokes if the flag was false.  (Defaulting to true means QE can check for its existence and fall back to the unload hack if it's not there.)
